storage_backend_ftp/components/ftp_adapter.py

```diff
@@ -0,0 +1,149 @@
+        ftp = ftplib.FTP_TLS
+        # Due to a bug into between ftplib and ssl, this part (about ssl) might not work!
+        # https://bugs.python.org/issue31727
+        security = None
+        if backend.ftp_security == "tls":
+            security = ssl.PROTOCOL_TLS
+        elif backend.ftp_security == "tlsv1":
+            security = ssl.PROTOCOL_TLSv1
+        elif backend.ftp_security == "tlsv1_1":
+            security = ssl.PROTOCOL_TLSv1_1
+        elif backend.ftp_security == "tlsv1_2":
+            security = ssl.PROTOCOL_TLSv1_2
+        elif backend.ftp_security == "sslv2":
+            security = ssl.PROTOCOL_SSLv2
+        elif backend.ftp_security == "sslv23":
+            security = ssl.PROTOCOL_SSLv23
+        elif backend.ftp_security == "sslv3":
+            security = ssl.PROTOCOL_SSLv3
+        if security:
+            ctx = ssl._create_stdlib_context(security)
+            params.update({"context": ctx})
+    else:
+        raise NotImplementedError()
+    with ftp(**params) as client:
+        client.connect(host=backend.ftp_server, port=backend.ftp_port)
+        if security:
+            client.auth()
+        client.login(backend.ftp_login, backend.ftp_password)
+        if security:
+            client.ssl_version = security
+            client.prot_p()
+        if backend.ftp_passive:
+            client.set_pasv(True)
+        yield client
+
+
+class FTPStorageBackendAdapter(Component):
+    _name = "ftp.adapter"
+    _inherit = "base.storage.adapter"
+    _usage = "ftp"
+
+    def add(self, relative_path, data, **kwargs):
+        with ftp(self.collection) as client:
+            full_path = self._fullpath(relative_path)
+            dirname = os.path.dirname(full_path)
+            if dirname:
+                try:
+                    client.cwd(dirname)
+                except IOError as e:
+                    if e.errno == errno.ENOENT:
+                        ftp_mkdirs(client, dirname)
+                    else:
+                        raise  # pragma: no cover
+            with io.BytesIO(data) as tmp_file:
+                try:
+                    client.storbinary("STOR " + full_path, tmp_file)
+                except ftplib.Error as e:
+                    raise ValueError(repr(e))
+                except OSError as e:
+                    raise ValueError(repr(e))
+
+    def get(self, relative_path, **kwargs):
+        full_path = self._fullpath(relative_path)
+        with ftp(self.collection) as client, StringIO() as buff:
+            try:
+                client.retrlines("RETR " + full_path, buff.write)
+            except ftplib.Error as e:
+                raise FileNotFoundError(repr(e))
+            buff.seek(0)
+            data = buff.read()
+        return data.encode()
+
+    def list(self, relative_path):
+        full_path = self._fullpath(relative_path)
+        with ftp(self.collection) as client:
+            try:
+                return client.retrlines("LIST " + full_path)
```

Comment:
This does not return a list. But a string with a command-like message. I think you should use `client.nlst(full_path)`

Reply:
addressed in https://github.com/acsone/storage/pull/1.